### PR TITLE
Witness Client side changes

### PR DESCRIPTION
**Pull request by yogeshbdeshpande**

### CODE REVIEW

**Comment by yogeshbdeshpande** [comment]
> e first thing I noticed from a quick scan is that it looks like the witness is now required. I think the witness should be optional, and something we can enable with flags. This allows the demo to be staged; we shouldn't _require_ the witness to give the existing basic demo.

Agree: The implementation makes it a Mandatory feature of verification. My impression was that STH Support will be an integral part of the demo, hence I built it as a mandatory feature.

**Comment by yogeshbdeshpande** [comment]
> > e first thing I noticed from a quick scan is that it looks like the witness is now required. I think the witness should be optional, and something we can enable with flags. This allows the demo to be staged; we shouldn't _require_ the witness to give the existing basic demo.
> 
> Agree: The implementation makes it a Mandatory feature of verification. My impression was that STH Support will be an integral part of the demo, hence I built it as a mandatory feature.

Thinking a bit more, I think, it makes it more logical to introduce it as a "Optional Feature".

As a result, I have modified the changes so that a Empty Witness URL passed to Flashing Tool skips the Witness checks.

If an explicit URL is introduced then only Witness Verification is attempted!

Please review the new version!

File: binary_transparency/firmware/internal/verify/bundle.go [hunk 1]
```diff
@@ -57,6 +57,39 @@ func BundleForUpdate(bundleRaw, fwHash []byte, dc api.LogCheckpoint, cpFunc Cons
 	return nil
 }
 
+// BundleValidateWitness verifies the received bundle against witness checkpoint
+func BundleValidateWitness(bundleRaw []byte, wc api.LogCheckpoint, cpFunc ConsistencyProofFunc) error {
+	var pb api.ProofBundle
+	lv := NewLogVerifier()
+	if err := json.Unmarshal(bundleRaw, &pb); err != nil {
+		return fmt.Errorf("failed to parse proof bundle: %w", err)
+	}
```
Comment (mhutchinson) [inline binary_transparency/firmware/internal/verify/bundle.go:66]: I'd be tempted to return the parsed ProofBundle from the method above, and then pass it back into this method if needed. Parsing it twice feels like wasted work, and just a little uncomfortable if the two parse methods ever diverged.

File: binary_transparency/firmware/internal/verify/bundle.go [hunk 2]
```diff
@@ -57,6 +57,39 @@ func BundleForUpdate(bundleRaw, fwHash []byte, dc api.LogCheckpoint, cpFunc Cons
...
+	fromts := wc.TreeSize
+	tots := pb.Checkpoint.TreeSize
+	root1 := wc.RootHash
+	root2 := pb.Checkpoint.RootHash
+	if wc.TreeSize > pb.Checkpoint.TreeSize {
+		fromts = pb.Checkpoint.TreeSize
+		tots = wc.TreeSize
+		root1 = pb.Checkpoint.RootHash
+		root2 = wc.RootHash
+	}
```
Comment (mhutchinson) [inline binary_transparency/firmware/internal/verify/bundle.go:82]: I think it's worth adding a couple of tests for the various orderings of:
 * bundle inclusion index
 * bundle cp
 * witness cp

I've checked the code here and it looks fine, but this block in particular would be easy to break during future changes.

**Review by mhutchinson — COMMENTED** [review]
Thanks Yogesh!

File: binary_transparency/firmware/cmd/flash_tool/flash_tool.go [hunk 3]
```diff
@@ -35,6 +35,7 @@ import (
 var (
 	deviceID      = flag.String("device", "", "One of [dummy, armory]")
 	logURL        = flag.String("log_url", "http://localhost:8000", "Base URL of the log HTTP API")
+	witnessURL    = flag.String("witness_url", "", "URL to fetch the Witness")
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply binary_transparency/firmware/cmd/flash_tool/flash_tool.go]: Changes Done!

File: binary_transparency/firmware/cmd/flash_tool/impl/flash_tool.go [hunk 4]
```diff
@@ -118,17 +135,12 @@ func readUpdateFile(path string) (api.UpdatePackage, error) {
...
-	// Get the consistency proof for the bundle
-	dc, err := dev.DeviceCheckpoint()
-	if err != nil {
-		return fmt.Errorf("failed to fetch the device checkpoint: %w", err)
-	}
-
+// getConsistencyFunc executes on a given client context and returns a
+// consistency function.
+func getConsistencyFunc(c *client.ReadonlyClient) func(from, to uint64) ([][]byte, error) {
 	cpFunc := func(from, to uint64) ([][]byte, error) {
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply binary_transparency/firmware/cmd/flash_tool/impl/flash_tool.go:141]: I would prefer this style as returning a function template appears a bit daunting to a reader, hence for the sake of simlicity of understanding, retained the original version!

File: binary_transparency/firmware/cmd/flash_tool/impl/flash_tool.go [hunk 5]
```diff
@@ -137,10 +149,39 @@ func verifyUpdate(c *client.ReadonlyClient, up api.UpdatePackage, dev devices.De
...
 	fwHash := sha512.Sum512(up.FirmwareImage)
 	if err := verify.BundleForUpdate(up.ProofBundle, fwHash[:], dc, cpFunc); err != nil {
 		return fmt.Errorf("failed to verify proof bundle: %w", err)
 	}
 	return nil
 }
+
+// verifyWitness checks that an update package is consistent with witness
+func verifyWitness(c *client.ReadonlyClient, wc client.WitnessClient, up api.UpdatePackage) error {
+
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply binary_transparency/firmware/cmd/flash_tool/impl/flash_tool.go]: Done!

File: binary_transparency/firmware/cmd/flash_tool/impl/flash_tool.go [hunk 6]
```diff
@@ -137,10 +149,39 @@ func verifyUpdate(c *client.ReadonlyClient, up api.UpdatePackage, dev devices.De
...
+	cpFunc := getConsistencyFunc(c)
 	fwHash := sha512.Sum512(up.FirmwareImage)
 	if err := verify.BundleForUpdate(up.ProofBundle, fwHash[:], dc, cpFunc); err != nil {
 		return fmt.Errorf("failed to verify proof bundle: %w", err)
 	}
 	return nil
 }
+
+// verifyWitness checks that an update package is consistent with witness
+func verifyWitness(c *client.ReadonlyClient, wc client.WitnessClient, up api.UpdatePackage) error {
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply binary_transparency/firmware/cmd/flash_tool/impl/flash_tool.go]: Yes, Agree! Done!

File: binary_transparency/firmware/internal/verify/bundle.go [hunk 7]
```diff
@@ -57,6 +57,39 @@ func BundleForUpdate(bundleRaw, fwHash []byte, dc api.LogCheckpoint, cpFunc Cons
...
 }
 
+// BundleValidateWitness verifies the received bundle against witness checkpoint
+func BundleValidateWitness(bundleRaw []byte, wc api.LogCheckpoint, cpFunc ConsistencyProofFunc) error {
+	var pb api.ProofBundle
+	lv := NewLogVerifier()
+	if err := json.Unmarshal(bundleRaw, &pb); err != nil {
+		return fmt.Errorf("failed to parse proof bundle: %w", err)
+	}
+	fmt.Printf("Witness TreeSize=%d, Inclusion Index=%d \n", wc.TreeSize, pb.InclusionProof.LeafIndex)
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply binary_transparency/firmware/internal/verify/bundle.go]: Done!

File: binary_transparency/firmware/internal/verify/bundle.go (hunk 1 above)
Comment (yogeshbdeshpande) [reply binary_transparency/firmware/internal/verify/bundle.go]: Agree, I have made the desired changes in the new revision, kindly check!

File: binary_transparency/firmware/internal/verify/bundle.go (hunk 2 above)
Comment (yogeshbdeshpande) [reply binary_transparency/firmware/internal/verify/bundle.go]: As discussed during F2F conversation, agreed to slightly simplify the coding style to better readability and understanding!
Uploaded changes with new implementation!

**Review by mhutchinson — APPROVED** [review]
Nice.

**Pull request merged**